src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -0,0 +1,1349 @@
+        // flow through to the terminal.
+        // GH#6423 - don't dismiss selection if the key that was pressed was a
+        // modifier key. We'll wait for a real keystroke to dismiss the
+        // GH #7395 - don't dismiss selection when taking PrintScreen
+        // selection.
+        // GH#8522, GH#3758 - Only dismiss the selection on key _down_. If we
+        // dismiss on key up, then there's chance that we'll immediately dismiss
+        // a selection created by an action bound to a keydown.
+        if (HasSelection() &&
+            !KeyEvent::IsModifierKey(vkey) &&
+            vkey != VK_SNAPSHOT &&
+            keyDown)
+        {
+            // GH#8791 - don't dismiss selection if Windows key was also pressed as a key-combination.
+            if (!eitherWinPressed)
+            {
+                _terminal->ClearSelection();
+                _renderer->TriggerSelection();
+            }
+
+            if (vkey == VK_ESCAPE)
+            {
+                return true;
+            }
+        }
+
+        // If the terminal translated the key, mark the event as handled.
+        // This will prevent the system from trying to get the character out
+        // of it and sending us a CharacterReceived event.
+        const auto handled = vkey ? _terminal->SendKeyEvent(vkey,
+                                                            scanCode,
+                                                            modifiers,
+                                                            keyDown) :
+                                    true;
+
+        return handled;
+    }
+
+    bool ControlCore::SendMouseEvent(const COORD viewportPos,
+                                     const unsigned int uiButton,
+                                     const ControlKeyStates states,
+                                     const short wheelDelta,
+                                     const ::Microsoft::Console::VirtualTerminal::TerminalInput::MouseButtonState state)
+    {
+        return _terminal->SendMouseEvent(viewportPos, uiButton, states, wheelDelta, state);
+    }
+    void ControlCore::UserScrollViewport(const int viewTop)
+    {
+        // Clear the regex pattern tree so the renderer does not try to render them while scrolling
+        _terminal->ClearPatternTree();
+
+        // This is a scroll event that wasn't initiated by the terminal
+        //      itself - it was initiated by the mouse wheel, or the scrollbar.
+        _terminal->UserScrollViewport(viewTop);
+    }
+
+    void ControlCore::AdjustOpacity(const double adjustment)
+    {
+        auto newOpacity = std::clamp(_settings.TintOpacity() + adjustment,
+                                     0.0,
+                                     1.0);
+        if (_settings.UseAcrylic())
+        {
+            try
+            {
+                _settings.TintOpacity(newOpacity);
+
+                if (newOpacity >= 1.0)

```

Comment:
you know, I literally was just bitten last night on a side project where I had a `curr == max` instead of a `curr >= max`, so I'm gonna leave it to be safe